src/freenet/support/io/InetAddressIpv6FirstComparator.java

```diff
@@ -17,80 +20,64 @@ public class InetAddressIpv6FirstComparator implements Comparator<InetAddress> {
-			return 1;
+		if(Objects.equals(arg0, arg1)) {
+			return 0;
 		}
-		// prefer everything over loopback / localhost
-		if (!arg0.isLoopbackAddress() && arg1.isLoopbackAddress()) {
-			return -1;
-		} else if (arg0.isLoopbackAddress() && !arg1.isLoopbackAddress()) {
-			return 1;
-		}
-		// prefer LAN routable addresses over link-local addresses because most of these will not work
-		if (!arg0.isLinkLocalAddress() && arg1.isLinkLocalAddress()) {
-			return -1;
-		} else if (arg0.isLinkLocalAddress() && !arg1.isLinkLocalAddress()) {
-			return 1;
-		}
-		// prefer reachable over unreachable addresses. This is usually a ping. TODO: This actually pings all advertised ip addresses. Is that OK? Do we need a maximum number of accepted addresses to prevent abuse as DDoS? 
-		int a = arg0.hashCode();
-		int b = arg1.hashCode();
-		Boolean reachable0 = reachabilityCache.get(a);
-		Boolean reachable1 = reachabilityCache.get(b);
-		if (reachable0 == null) {
-			try {
-				reachable0 = arg0.isReachable((int) DEFAULT_MAX_PING_TIME);
-			} catch (IOException e) {
-				reachable0 = false;
-			}
-			reachabilityCache.put(a, reachable0);
-		}
-		if (reachable1 == null) {
+		return innerComparator.compare(arg0, arg1);
+	}
+
+	private boolean isReachable(InetAddress inetAddress) {
+		int hashCode = inetAddress.hashCode();
+		Boolean reachable = reachabilityCache.get(hashCode);
+		if (reachable == null) {
 			try {
-				reachable1 = arg1.isReachable((int) DEFAULT_MAX_PING_TIME);
+				reachable = inetAddress.isReachable((int) DEFAULT_MAX_PING_TIME);
 			} catch (IOException e) {
-				reachable1 = false;
+				reachable = false;
 			}
-			reachabilityCache.put(b, reachable1);
-		}
-		if (reachable0 && !reachable1) {
-			return -1;
-		} else if (!reachable0 && reachable1) {
-			return 1;
-		}
-		// among routable addresses prefer LAN addresses over global addresses, because they should stay within VPNs
-		if (!arg0.isSiteLocalAddress() && arg1.isSiteLocalAddress()) {
-			return -1;
-		} else if (arg0.isSiteLocalAddress() && !arg1.isSiteLocalAddress()) {
-			return 1;
-		}
-		byte[] bytes0 = arg0.getAddress();
-		byte[] bytes1 = arg1.getAddress();
-		// prefer IPv6 over IPv4
-		if(bytes0.length > bytes1.length) {
-			return -1;
-		} else if(bytes1.length > bytes0.length) {
-			return 1;
+			reachabilityCache.put(hashCode, reachable);
 		}
+		return reachable;
+	}
+
+	// inverted prefer, because not(...) is only documented since Java 11
+	private Comparator<InetAddress> preferNot(Predicate<InetAddress> pred) {
+		return (arg0, arg1) -> -1 * predicateToCompare(pred, arg0, arg1);
+	}
 
-		// Sort by hash code as fallback. This is fast.
-		if(a > b) return 1;
-		else if(b > a) return -1;
-		return Fields.compareBytes(bytes0, bytes1);
-		// Hostnames in InetAddress are merely cached, equals() only operates on the byte[].
+	private Comparator<InetAddress> prefer(Predicate<InetAddress> pred) {
+		return (arg0, arg1) -> predicateToCompare(pred, arg0, arg1);
```

Comment:
This can be simplified to `return (arg0, arg1) -> Boolean.compare(!pred.test(arg0), !pred.test(arg1));`

Also avoids having to test the predicates twice.